src/uv_tcp_connect.c

```diff
@@ -146,55 +155,111 @@ static void uvTcpConnectUvConnectCb(struct uv_connect_s *req, int status)
     uvTcpConnectAbort(connect);
 }
 
-/* Create a new TCP handle and submit a connection request to the event loop. */
-static int uvTcpConnectStart(struct uvTcpConnect *r, const char *address)
+/* The hostname resolve is finished */
+static void uvGetAddrInfoCb(uv_getaddrinfo_t *req,
+                            int status,
+                            struct addrinfo *res)
 {
-    struct UvTcp *t = r->t;
-    struct sockaddr_in addr;
+    struct uvTcpConnect *connect = req->data;
+    struct UvTcp *t = connect->t;
     int rv;
 
-    rv = uvIpParse(address, &addr);
+    connect->resolving =
+        false; /* Indicate we are in the name resolving phase */
+
+    if (t->closing) {
+        connect->status = RAFT_CANCELED;
+
+        /* We need to close the tcp handle to to connection attempt */
+        uv_close((struct uv_handle_s *)connect->tcp, uvTcpConnectUvCloseCb);
+        return;
+    }
+
+    if (status < 0) {
+        ErrMsgPrintf(t->transport->errmsg, "uv_getaddrinfo(): %s",
+                     uv_err_name(status));
+        connect->status = RAFT_NOCONNECTION;
+        goto err;
+    }
+    rv = uv_tcp_connect(&connect->connect, connect->tcp,
+                        (const struct sockaddr *)res->ai_addr,
+                        uvTcpConnectUvConnectCb);
     if (rv != 0) {
+        /* UNTESTED: since parsing succeed, this should fail only because of
+         * lack of system resources */
+        ErrMsgPrintf(t->transport->errmsg, "uv_tcp_connect(): %s",
+                     uv_strerror(rv));
+        connect->status = RAFT_NOCONNECTION;
         goto err;
     }
 
+    return;
+
+err:
+    uvTcpConnectAbort(connect);
+}
+/* Create a new TCP handle and submit a connection request to the event loop. */
+static int uvTcpConnectStart(struct uvTcpConnect *r, const char *address)
+{
+    static struct addrinfo hints = {
+        .ai_flags = AI_V4MAPPED | AI_ADDRCONFIG | AI_NUMERICHOST,
+        .ai_family = AF_INET,
+        .ai_socktype = SOCK_STREAM,
+        .ai_protocol = 0};
+    struct UvTcp *t = r->t;
+    char hostname[NI_MAXHOST];
+    char service[NI_MAXSERV];
+    int rv;
+
+    r->handshake.base = NULL;
+
     /* Initialize the handshake buffer. */
     rv = uvTcpEncodeHandshake(t->id, t->address, &r->handshake);
     if (rv != 0) {
         assert(rv == RAFT_NOMEM);
-        ErrMsgOom(r->t->transport->errmsg);
+        ErrMsgOom(t->transport->errmsg);
         goto err;
```

Comment:
`handshake.base` is freed after `err`, but never allocated. It looks safe right now, but it's not very pretty and can lead to an erroneous free in the future.

Reply:
I initialise the pointer to NULL in the beginning of the function. As the raft_free (like the posix free is safe to be invoked on a NULL pointer) it safe from my perspective.

Reply:
It's just the default port if none is specified. We should indeed document it, or perhaps return an error instead.